util/chunk/chunk.go

```diff
@@ -112,12 +157,68 @@ func (c *Chunk) SetNumVirtualRows(numVirtualRows int) {
 // Reset resets the chunk, so the memory it allocated can be reused.
 // Make sure all the data in the chunk is not used anymore before you reuse this chunk.
 func (c *Chunk) Reset() {
-	for _, c := range c.columns {
-		c.reset()
+	for _, col := range c.columns {
+		col.reset()
 	}
 	c.numVirtualRows = 0
 }
 
+// GrowAndReset resets the chunk and grow capacity if needed.
+// TODO: this method will be used in continue PR.
+func (c *Chunk) GrowAndReset(maxChunkSize int) {
+	if c.columns == nil {
+		return
+	}
+	needRenewCol, newCap := calculateCapForReset(c, maxChunkSize)
+	if !needRenewCol {
+		c.Reset()
+		return
+	}
+	c.capacity = newCap
+	c.columns = renewColumns(c.columns, newCap)
+	c.numVirtualRows = 0
+}
+
+// calculateCapForReset checks and return whether need scale up and suitable capacity execution.
+//  It works like:
+//      keep old value if capacity == maxChunkSize or chk isn't full, and no need renew columns.
+//      old value x 2  if chk is full, and need renew columns.
+func calculateCapForReset(c *Chunk, maxChunkSize int) (requireRenewCol bool, newCap int) {
+	if c.capacity == maxChunkSize || c.NumRows() < c.capacity {
+		return false, c.capacity
+	}
+	return true, mathutil.Min(c.capacity*2, maxChunkSize)
+}
+
+// calculateCapForRenew checks and return suitable capacity in next execution.
+//  It works like:
+//      using next power of 2 for renew if currSize < cap
+//      using capacity limit if full but capacity already meet maxChunkSize
+//      using oldcap x 2 if chk is full but less than maxChunkSize
+func calculateCapForRenew(c *Chunk, maxChunkSize int) int {
+	currSize := c.NumRows()
+	if currSize < c.capacity {
+		return NextPowerOfTwo(currSize)
+	}
+	if c.capacity == maxChunkSize {
+		return maxChunkSize
+	}
+	return mathutil.Min(c.capacity*2, maxChunkSize)
```

Comment:
Can we remove line 203~205? IMO, `mathutil.Min(c.capacity*2, maxChunkSize)` can ccover the case of `c.capacity == maxChunkSize`.